### PR TITLE
[wip] add emp-ssl

### PR DESCRIPTION
**Pull request by kumarkrishna**
- [x] add EMP-SSL
- [x] add wandb logging
- [x] write multi-view datasets

### CODE REVIEW

File: fastssl/models/emp.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+def EMPLoss(model, inp, _lambda=None, patch_sim_scale=200, num_patches=2):
+    """
+    Implement Extreme Multi-Patch training.
+    """
+    # build losses
+    simloss = ClusterCosineSimilarity()
+    tcrloss = TotalCodingRate()
+
+    inp_multiview = torch.cat(inp, dim=0)
```
Comment (arnaghosh) [inline fastssl/models/emp.py:68]: Interesting that the effective batchsize for forward propagation now becomes num_patches x bsz?

File: fastssl/utils/base.py [hunk 2]
```diff
@@ -2,9 +2,11 @@
 import random, os
 import torch
 import numpy as np
+import os
 
 from fastargs import get_current_config
 
+
```
Comment (arnaghosh) [inline fastssl/utils/base.py:9]: This file is probably behind main. Merging will be required.

File: fastssl/models/vicreg.py
Comment (arnaghosh) [inline fastssl/models/vicreg.py:1]: Remove file from this PR?

File: mila_cc_job_scripts/wandb_sweep_script.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+export CHECKPOINT_DIR=/home/mila/a/arnab.mondal/scratch/fastssl/checkpoints
+
+wandb agent eigengroup/fastssl/54megz1q
+
+# export FFCV_DATA_DIR=/network/projects/_groups/linclab_users/ffcv/ffcv_datasets
+# export TORCHVISION_DATA_DIR=/network/datasets
+# export CHECKPOINT_DIR=/network/projects/_groups/linclab_users/fastssl/checkpoints
+
+
+#wandb sweep --project fastssl -e eigengroup wandb_sweep_configs/design_hyperparams_sweep.yaml
```
Comment (arnaghosh) [inline mila_cc_job_scripts/wandb_sweep_script.sh:28]: Possibly add some comments to this file to remind ourselves what/how to run the wandb sweeps?

File: python_scripts/run_wandb.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ssl_train_cmd = 'python python_scripts/train_model.py'
+for arg in all_args:
+    ssl_train_cmd += ' {}'.format(arg)
+    if 'algorithm' in arg:
+        assert arg.split('=')[1] in config_files_dict.keys(), "SSL Algorithm {} not implemented".format(arg.split('=')[1]) 
+        ssl_train_cmd += ' --config-file configs/cc_{}.yaml'.format(arg.split('=')[1])
+print(ssl_train_cmd)
+os.system(ssl_train_cmd)
+
+for seed in seeds:
```
Comment (arnaghosh) [inline python_scripts/run_wandb.py:25]: This probably needs to be changed now because we want to run 3 seeds of SSL pretraining and corresponding linear eval with the same seed. Hence, put the `ssl_train_cmd` part (L16-L23) inside this loop.
Ok for now, but make change in future PR.

File: wandb_sweep_configs/design_hyperparams_sweep_BarlowTwins.yaml
Comment (arnaghosh) [inline wandb_sweep_configs/design_hyperparams_sweep_BarlowTwins.yaml:1]: Remove from PR?

File: python_scripts/train_model.py
Comment (arnaghosh) [inline python_scripts/train_model.py:1]: This probably lacks the precaching and model.backbone.eval() changes (present in main). 
Check before merging with main.

**Review by arnaghosh — APPROVED** [review]
Approved with comments. Merging might be tricky though. Please be careful of the changes in main.
